rust/worker/src/system/types.rs

```diff
@@ -72,17 +74,96 @@ where
     }
 }
 
+/// A thin wrapper over a join handle that will panic if it is consumed twice.
+#[derive(Debug)]
+struct ConsumableJoinHandle {
+    handle: Option<tokio::task::JoinHandle<()>>,
+}
+
+impl ConsumableJoinHandle {
+    fn new(handle: tokio::task::JoinHandle<()>) -> Self {
+        ConsumableJoinHandle {
+            handle: Some(handle),
+        }
+    }
+
+    async fn consume(&mut self) -> Result<(), JoinError> {
+        match self.handle.take() {
+            Some(handle) => {
+                handle.await?;
+                Ok(())
+            }
+            None => {
+                panic!("Join handle already consumed");
+            }
+        }
+    }
+}
+
+/// A ComponentSender is generic over a component type. This struct is internal to the system module.
+/// It's implemented as a struct instead of a type alias so that it can implement common logic around the channel.
+///
+/// See ReceiverForMessage for a trait generic over a message type.
+#[derive(Debug)]
+pub(super) struct ComponentSender<C: Component> {
+    sender: tokio::sync::mpsc::Sender<WrappedMessage<C>>,
+}
+
+impl<C: Component> ComponentSender<C> {
+    pub(super) fn new(sender: tokio::sync::mpsc::Sender<WrappedMessage<C>>) -> Self {
+        ComponentSender { sender }
+    }
+
+    pub(super) async fn wrap_and_send<M>(
+        &self,
+        message: M,
+        tracing_context: Option<tracing::Span>,
+    ) -> Result<(), ChannelError>
+    where
+        C: Handler<M>,
+        M: Send + Debug + 'static,
+    {
+        self.sender
+            .send(wrap(message, tracing_context))
+            .await
+            .map_err(|_| ChannelError::SendError)
+    }
+}
+
+// Cannot automatically derive, see https://github.com/rust-lang/rust/issues/26925
+impl<C: Component> Clone for ComponentSender<C> {
+    fn clone(&self) -> Self {
+        ComponentSender {
+            sender: self.sender.clone(),
+        }
+    }
+}
+
 /// A component handle is a handle to a component that can be used to stop it.
 /// and introspect its state.
 /// # Fields
 /// - cancellation_token: A cancellation token that can be used to stop the component
 /// - state: The state of the component
 /// - join_handle: The join handle for the component, used to join on the component
-pub(crate) struct ComponentHandle<C: Component> {
+/// - sender: A channel to send messages to the component
+#[derive(Debug)]
+pub(crate) struct ComponentHandle<C: Component + Debug> {
     cancellation_token: tokio_util::sync::CancellationToken,
-    state: ComponentState,
-    join_handle: Option<tokio::task::JoinHandle<()>>,
-    sender: Sender<C>,
+    state: Arc<Mutex<ComponentState>>,
```

Comment:
I don't think we need this to be tokio::mutex? and could instead do a parking_lot::mutex as is standard across our codebase